std-bits/table/src/main/java/org/enso/table/data/column/builder/InferredIntegerBuilder.java

```diff
@@ -0,0 +1,115 @@
+package org.enso.table.data.column.builder;
+
+import java.math.BigInteger;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.column.storage.type.BigIntegerType;
+import org.enso.table.data.column.storage.type.IntegerType;
+import org.enso.table.data.column.storage.type.StorageType;
+import org.enso.table.problems.ProblemAggregator;
+
+/**
+ * A builder for storing enso Integers, which might be Longs or BigIntegers.
+ *
+ * <p>This builder starts off delegating to LongBuilder, but if it receives a BigInteger, it retypes
+ * the LongBuilder to a BigIntegerBuilder.
+ */
+public class InferredIntegerBuilder extends Builder {
+  private LongBuilder longBuilder = null;
+  private TypedBuilder bigIntegerBuilder = null;
+  private int currentSize = 0;
+  private final int initialSize;
+  private final ProblemAggregator problemAggregator;
+
+  /** Creates a new instance of this builder, with the given known result length. */
+  public InferredIntegerBuilder(int initialSize, ProblemAggregator problemAggregator) {
+    this.initialSize = initialSize;
+    this.problemAggregator = problemAggregator;
+
+    longBuilder =
+        NumericBuilder.createLongBuilder(this.initialSize, IntegerType.INT_64, problemAggregator);
+  }
+
+  @Override
+  public void appendNoGrow(Object o) {
+    if (o == null) {
+      appendNulls(1);
+    } else if (o instanceof BigInteger bi) {
+      retypeToBigIntegerMaybe();
+      bigIntegerBuilder.appendNoGrow(bi);
+    } else if (o instanceof Long lng) {
+      if (bigIntegerBuilder != null) {
+        bigIntegerBuilder.appendNoGrow(BigInteger.valueOf(lng));
+      } else {
+        longBuilder.appendNoGrow(lng);
+      }
+    }
+    currentSize++;
+  }
+
+  @Override
+  public void append(Object o) {
+    if (o == null) {
+      appendNulls(1);
+    } else if (o instanceof BigInteger bi) {
+      retypeToBigIntegerMaybe();
+      bigIntegerBuilder.append(bi);
+    } else if (o instanceof Long lng) {
+      if (bigIntegerBuilder != null) {
+        bigIntegerBuilder.append(BigInteger.valueOf(lng));
+      } else {
+        longBuilder.append(lng);
+      }
+    }
```

Comment:
Note that in the `LongBuilder*` classes this is rather done by: 
```
Long x = NumericConverter.tryConvertingToLong(o);
```

That's because it was possible in some cases (not sure exactly which tbh) to have `Short` or other Java integer types flying around. And more importantly, because adding `2.0` to an integer builder used to _not_ change its type to decimal - it was kept as integers if the floating point value was a whole integer.

Reply:
Also I see no `else` branch here.

What happens if someone tries to append a non-numeric value here? It looks that `currentSize` is incremented but the underlying builder is not really updated. That seems to break the 'synchronization' between the `currentSize` and the underlying builder. Note that in `InferredBuilder` every increment of `currentSize` is associated with _some_ append operation on the `currentBuilder` - the only exception is if `currentBuilder == null` - but then we explicitly ensure the synchronization in `initBuilderFor` by the line `currentBuilder.appendNulls(currentSize);`. I imagine this builder would need a similar logic.

Reply:
There is a lot of edge cases here, `InferredBuilder` had a few bugs related to this complexity. If we are adding a separate builder that has similar but differing logic, I think it would be good to add some more tests for edge cases in cases where this builder is being used - e.g. starting with many `Nothing` results first and only later getting actual numbers, checking what happens if a few values are long and only later we get a big integer etc.

Reply:
> Note that in the `LongBuilder*` classes this is rather done by:
> 
> ```
> Long x = NumericConverter.tryConvertingToLong(o);
> ```
> 
> That's because it was possible in some cases (not sure exactly which tbh) to have `Short` or other Java integer types flying around. And more importantly, because adding `2.0` to an integer builder used to _not_ change its type to decimal - it was kept as integers if the floating point value was a whole integer.

As for this comment - I realize that this may not be necessary here if the builder is used only for the round operations and we assume it only takes Long or BigInteger. But then we should throw an exception if an unexpected value is appended - so that if anyone violates this assumption they will know this builder does not support it / would need to be extended.

Reply:
I changed it to use `tryConvertingToLong` and throw an exception if the type is wrong.